### DISCUSSION TITLE
How to implement authorization in the best way?

### DISCUSSION THREAD

**Discussion by jakubjereczek** [vercel/next.js#50674] [App Router]
Hi guys,

I found many examples of how to implement authorization using external libraries like NextAuth, Auth0 or AuthSignal but but what are the recommended steps to implement custom authorization?
I have planned to migrate application from React into Next 13. 

I suppose the best way to handle this in Server Component is to put Authorization header. 
In the case of HTTP request, when a request returns a 401/403 error status and token already expired we should return a response to the component indicating a lack of authorization (NextResponse?).

Could I ask for some tips on how to do it best? It mainly concerns how to handle it on the server-side components.

**Comment by jakubjereczek** [comment]
bump

**Reply by tinleym** [reply]
I haven't been able to figure a practical way to do this without framework support.

Ultimately, I think the thing for Next to do is to support 401/403s the same way as it does 404s... in the case of 403s, have a forbidden() server function like notFound() and a forbidden.jsx file like not-found.jsx. 

In the meantime, I'm just treating 401/403s as 404s where I can. I haven't figured out how to best route a resource like /[id]/privileged that might not exist and/or be forbidden to the client.

Thinking about it more, clients can be forbidden access for different reasons (e.g. not logged in, blocked)... so it would useful to have a reason parameter to a forbidden() function.
